pkg/controller/direct/firestore/firestoredocument_mappers.go

```diff
@@ -35,16 +39,123 @@ func FirestoreDocumentSpec_v1alpha1_FromProto(mapCtx *direct.MapContext, in *pb.
 	for k, v := range in.Fields {
 		outV := Field_FromProto(mapCtx, v)
 
-		j, err := json.Marshal(outV)
+		j, err := toJSON(outV)
 		if err != nil {
-			return nil
+			mapCtx.Errorf("failed to marshal field in FirestoreDocument: %v", err)
+			continue
 		}
-		out.Fields[k] = apiextensionsv1.JSON{Raw: j}
+		out.Fields[k] = j
 	}
 
 	return out
 }
 
+// toJSON converts a Go value to an apiextensionsv1.JSON representation,
+// taking care to preserve integer and float types to avoid type conversion issues,
+// even when nested in an array or map.
+func toJSON(in any) (apiextensionsv1.JSON, error) {
+	var buffer bytes.Buffer
+
+	render := in
+	switch in := in.(type) {
+	case []any:
+		arr := make([]apiextensionsv1.JSON, 0, len(in))
+		for _, elem := range in {
+			v, err := toJSON(elem)
+			if err != nil {
+				return apiextensionsv1.JSON{}, err
+			}
+			arr = append(arr, v)
+		}
+		render = arr
+	case map[string]any:
+		m := make(map[string]apiextensionsv1.JSON)
+		for k, elem := range in {
+			v, err := toJSON(elem)
+			if err != nil {
+				return apiextensionsv1.JSON{}, err
+			}
+			m[k] = v
+		}
+		render = m
+
+	case int, int32, int64:
+		// As a special case, we want to avoid encoding int64 as a float in JSON,
+		// because this causes type conversions.
+		// So we manually format it as a string.
+		var buffer bytes.Buffer
+		fmt.Fprintf(&buffer, "%d", in)
+		render = buffer.String()
+		return apiextensionsv1.JSON{Raw: buffer.Bytes()}, nil
+	case float32, float64:
+		// Force float to be formatted with scientific notation,
+		// so we don't convert back to int accidentally.
+		var buffer bytes.Buffer
+		fmt.Fprintf(&buffer, "%g", in)
+		render = buffer.String()
+		return apiextensionsv1.JSON{Raw: buffer.Bytes()}, nil
+	}
```

Comment:
Are we deliberately letting unrecognized types fall through? Even if we are I think a 'default:' with a comment that says that is easier to understand.